File: selfdoc_core/catalog.py
"""Directive catalog: defines all built-in directive names and their status.

The ``CORE_DIRECTIVES`` catalogue is no longer a hand-maintained dict literal.
It is BUILT at import time from ``selfdoc_core/directives.toml`` -- a declarative
descriptor document governed by ``.strictspec/directive-descriptor.schema.toml``
and validated by the strictspec-generated ``directive_descriptor_validator``.
The document is the single source of truth; this module is a thin loader.

A malformed catalogue document (bad name grammar, unknown key, missing required
field, duplicate name, absent format_version gate, ...) is a hard error at import
-- selfdoc fails to load before any directive is dispatched.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from importlib.resources import files

from selfdoc_core import directive_descriptor_validator as _validator


@dataclass(slots=True)
class DirectiveSpec:
    """Metadata for a single built-in directive."""

    description: str
    category: str  # "code" or "content"
    required_attrs: list[str] = field(default_factory=list)
    optional_attrs: list[str] = field(default_factory=list)
    example: str = ""


# -- Shared cross-cutting attributes ------------------------------------------

# Attributes accepted by every code-category directive regardless of name.
# The multi-language resolver (selfdoc_core.resolver.Resolver) reads ``lang``
# to disambiguate which language extractor handles a path-dispatched directive,
# so it is valid on any code directive. gen emits it on every generated ``ref``
# page. Each code directive's ``optional_attrs`` in ``directives.toml`` lists it
# explicitly; this constant names the invariant a test enforces.
SHARED_CODE_ATTRS: list[str] = ["lang"]


# -- Core directives (built from the validated descriptor document) -----------

_CATALOGUE_DOCUMENT = "directives.toml"


class CatalogDocumentError(RuntimeError):
    """Raised when ``directives.toml`` fails strictspec validation at load.

    A hard error at import time: the built-in directive catalogue is malformed,
    so selfdoc cannot know what its own directives are. Subclasses
    ``RuntimeError`` so the CLI's existing handlers surface it cleanly.
    """


def _build_catalogue(raw: bytes) -> dict[str, DirectiveSpec]:
    """Validate raw catalogue-document bytes and bind them into DirectiveSpecs.

    strictspec is the boundary validator: the document is validated against its
    schema by the generated validator, and only a fully-valid document is bound.
    Any diagnostic is a hard error (:class:`CatalogDocumentError`).
    """
    catalogue, diags = _validator.validate_bytes(raw, "toml")
    if diags:
        detail = "\n".join(f"  {d.path}: {d.message} [{d.code}]" for d in diags)
        raise CatalogDocumentError(
            f"{_CATALOGUE_DOCUMENT} is not a valid directive catalogue:\n{detail}"
        )
    return {
        d.name: DirectiveSpec(
            description=d.description,
            category=d.category,
            required_attrs=list(d.required_attrs),
            optional_attrs=list(d.optional_attrs),
            example=d.example,
        )
        for d in catalogue.directives
    }


def _load_core_directives() -> dict[str, DirectiveSpec]:
    """Read, validate, and bind ``directives.toml`` into the runtime catalogue."""
    raw = files("selfdoc_core").joinpath(_CATALOGUE_DOCUMENT).read_bytes()
    return _build_catalogue(raw)


CORE_DIRECTIVES: dict[str, DirectiveSpec] = _load_core_directives()

# -- Future directives (declared, not yet implemented) ------------------------

FUTURE_DIRECTIVES: set[str] = {
    # Tables
    "table-param",
    "table-env",
    "table-compare",
    "table-error",
    "table-shortcut",
    "table-status",
    "table-registry",
    "table-migration",
    "table-timeline",
    "table-perm",
    "table-plan",
    # Code
    "code-source",
    "code-example",
    "code-session",
    "code-repl",
    "code-config",
    "code-diff",
    "code-error",
    "code-schema",
    "code-template",
    "code-log",
    "code-query",
    "code-wire",
    "code-build",
    # Lists
    "list-toc",
    "list-check",
    "list-steps",
    "list-faq",
    "list-deps",
    "list-breadcrumb",
    "list-related",
    "list-errors",
    "list-decisions",
    "list-reqs",
    "list-api",
    "list-changelog",
    # Callouts
    "callout-example",
    "callout-deprecated",
    "callout-security",
    "callout-perf",
    "callout-compat",
    "callout-experimental",
    "callout-see-also",
    "callout-breaking",
    "callout-success",
    "callout-quote",
    # Prose
    "prose-summary",
    "prose-caption",
    "prose-rationale",
    "prose-caveat",
    "prose-migration",
    "prose-changelog",
    "prose-release",
    "prose-prereq",
    "prose-abstract",
    "prose-deprecation",
    "prose-attribution",
    "prose-definition",
    "prose-annotation",
    "prose-example",
}

# -- Combined set of all built-in directives ----------------------------------

ALL_BUILTIN_DIRECTIVES: set[str] = set(CORE_DIRECTIVES) | FUTURE_DIRECTIVES

# -- Old-to-new name mapping for backward compatibility -----------------------

DIRECTIVE_NAME_MAPPING: dict[str, str] = {
    "module": "ref",
    "schema": "table-schema",
    "test": "code-test",
    "cli": "code-help",
    "config": "table-config",
    "glossary": "list-glossary",
}


# -- Helper functions ---------------------------------------------------------


def is_valid_directive(
    name: str, custom_names: set[str] | None = None
) -> bool:
    """Return True if *name* is a recognized built-in or custom directive."""
    if name in ALL_BUILTIN_DIRECTIVES:
        return True
    if custom_names is not None and name in custom_names:
        return True
    return False


class DirectiveAttrError(RuntimeError):
    """Raised when a directive uses an unknown attribute or omits a required one.

    This is a hard error (exit 1), distinct from resolution failures which are
    warning-level. It subclasses ``RuntimeError`` so the CLI's existing
    ``RuntimeError`` handlers surface it with a clean message and a non-zero
    exit code.
    """


def validate_directive_attrs(
    name: str, attrs: dict[str, str], *, file: str, line: int
) -> None:
    """Enforce a directive's attribute contract against its catalog spec.

    Raises :class:`DirectiveAttrError` if *attrs* contains an attribute the
    directive does not accept, or omits one it requires. Only core directives
    with a known spec are enforced; custom and future directives are skipped
    because they define their own attribute schemas.

    Args:
        name: Directive name (e.g. ``"ref"``).
        attrs: Parsed attribute dict for this directive occurrence.
        file: Source file path, for the error message.
        line: 1-based line number within *file*, for the error message.
    """
    spec = CORE_DIRECTIVES.get(name)
    if spec is None:
        return

    allowed = set(spec.required_attrs) | set(spec.optional_attrs)

    for attr in attrs:
        if attr in allowed:
            continue
        # Actionable migration hint for the removed table-commands path attr.
        if name == "table-commands" and attr == "path":
            raise DirectiveAttrError(
                f"{file}:{line}: directive 'table-commands' no longer takes "
                "'path'; the schema is discovered automatically. Use "
                'schema-dir="<dir>" only if discovery reports ambiguity.'
            )
        allowed_display = ", ".join(sorted(allowed)) if allowed else "(none)"
        raise DirectiveAttrError(
            f"{file}:{line}: directive '{name}' has unknown attribute "
            f"'{attr}'. Allowed attributes: {allowed_display}."
        )

    for req in spec.required_attrs:
        if req not in attrs:
            required_display = ", ".join(spec.required_attrs)
            raise DirectiveAttrError(
                f"{file}:{line}: directive '{name}' is missing required "
                f"attribute '{req}'. Required attributes: {required_display}."
            )


def directive_status(name: str) -> str:
    """Return ``"core"``, ``"future"``, or ``"unknown"`` for a directive name.

    Only checks built-in directives; custom directives are the caller's
    responsibility.
    """
    if name in CORE_DIRECTIVES:
        return "core"
    if name in FUTURE_DIRECTIVES:
        return "future"
    return "unknown"

File: internal/catalog/catalog.go
// Package catalog is selfdoc's directive catalogue: every built-in directive
// name and its status.
//
// The shipped ("core") catalogue is not a hand-maintained literal. It is built
// from the embedded directives.toml -- a declarative descriptor document
// governed by .strictspec/directive-descriptor.schema.toml and validated by the
// strictspec-generated validator in this package. The document is the single
// authority; this file is a thin loader.
//
// A malformed catalogue document (bad name grammar, unknown key, missing
// required field, duplicate name, absent format_version marker) is a hard
// error before any directive is dispatched. [Load] reports it as an error and
// [Core] panics, which is the Go counterpart of the Python surface's
// import-time crash: the document is compiled into the binary, so a panic here
// means the binary itself is malformed.
package catalog

import (
	_ "embed"
	"fmt"
	"sort"
	"strings"
	"sync"
)

// catalogueDocumentName is the document's name as the error messages spell it.
const catalogueDocumentName = "directives.toml"

//go:embed directives.toml
var catalogueDocument []byte

// DirectiveSpec is the metadata for a single built-in directive.
type DirectiveSpec struct {
	// Description is the one-line human description the directive reference
	// table renders.
	Description string
	// Category is "code" (dispatches to a language extractor by path) or
	// "content" (resolves language-agnostically).
	Category string
	// RequiredAttrs and OptionalAttrs are the directive's attribute contract,
	// in document order.
	RequiredAttrs []string
	OptionalAttrs []string
	// Example is a representative usage example.
	Example string
}

// SharedCodeAttrs are the attributes every code-category directive accepts
// regardless of name.
//
// The multi-language resolver reads "lang" to decide which extractor handles a
// path-dispatched directive, and gen emits it on every generated ref page.
// Each code directive's optional_attrs in directives.toml lists it explicitly;
// this is the invariant a test enforces, not a second declaration.
func SharedCodeAttrs() []string { return []string{"lang"} }

// CatalogDocumentError reports a catalogue document that failed strictspec
// validation: the built-in directive catalogue is malformed, so selfdoc cannot
// know what its own directives are.
type CatalogDocumentError struct {
	Message string
}

func (e *CatalogDocumentError) Error() string { return e.Message }

// Catalogue is a loaded, validated directive catalogue.
//
// It keeps the document's order, because the directive reference table renders
// in it, and answers a name lookup in constant time.
type Catalogue struct {
	order []string
	specs map[string]DirectiveSpec
}

// Names returns every catalogued directive name in document order.
func (c *Catalogue) Names() []string {
	out := make([]string, len(c.order))
	copy(out, c.order)
	return out
}

// Len returns how many directives the catalogue carries.
func (c *Catalogue) Len() int { return len(c.order) }

// Spec returns the named directive's metadata, reporting whether the
// catalogue carries it.
func (c *Catalogue) Spec(name string) (DirectiveSpec, bool) {
	spec, ok := c.specs[name]
	return spec, ok
}

// Has reports whether the catalogue carries name.
func (c *Catalogue) Has(name string) bool {
	_, ok := c.specs[name]
	return ok
}

// BuildCatalogue validates raw catalogue-document bytes and binds them into a
// [Catalogue].
//
// strictspec is the boundary validator: the document is checked against its
// schema by the generated validator, and only a wholly valid document is
// bound. Any diagnostic is a [CatalogDocumentError] -- never a silent partial
// catalogue.
func BuildCatalogue(raw []byte) (*Catalogue, error) {
	document, diags := ValidateBytes(raw, "toml")
	if len(diags) > 0 {
		var detail strings.Builder
		for _, d := range diags {
			fmt.Fprintf(&detail, "\n  %s: %s [%s]", d.Path, d.Message, d.Code)
		}
		return nil, &CatalogDocumentError{Message: fmt.Sprintf(
			"%s is not a valid directive catalogue:%s",
			catalogueDocumentName, detail.String())}
	}
	c := &Catalogue{
		order: make([]string, 0, len(document.Directives)),
		specs: make(map[string]DirectiveSpec, len(document.Directives)),
	}
	for _, d := range document.Directives {
		c.order = append(c.order, d.Name)
		c.specs[d.Name] = DirectiveSpec{
			Description:   d.Description,
			Category:      d.Category,
			RequiredAttrs: append([]string(nil), d.RequiredAttrs...),
			OptionalAttrs: append([]string(nil), d.OptionalAttrs...),
			Example:       d.Example,
		}
	}
	return c, nil
}

// Load reads, validates and binds the embedded catalogue document.
//
// This is the error-returning door. Production code reads [Core] instead,
// which loads once.
func Load() (*Catalogue, error) { return BuildCatalogue(catalogueDocument) }

var (
	coreOnce sync.Once
	core     *Catalogue
	coreErr  error
)

// Core returns the shipped catalogue, loading and validating the embedded
// document on first use.
//
// It panics when the document is malformed. That is not a judgement call: the
// document is embedded in the binary, so a diagnostic here means this build of
// selfdoc does not know what its own directives are, and every caller below
// would have to invent a behavior for a catalogue that cannot exist. Use
// [Load] where an error is wanted.
func Core() *Catalogue {
	coreOnce.Do(func() { core, coreErr = Load() })
	if coreErr != nil {
		panic(coreErr)
	}
	return core
}

// futureDirectives are the declared-but-unimplemented directive names.
//
// They are recognized so a document may name one without failing validation,
// and skipped by attribute enforcement because they carry no descriptor yet.
var futureDirectives = map[string]struct{}{
	// Tables
	"table-param":     {},
	"table-env":       {},
	"table-compare":   {},
	"table-error":     {},
	"table-shortcut":  {},
	"table-status":    {},
	"table-registry":  {},
	"table-migration": {},
	"table-timeline":  {},
	"table-perm":      {},
	"table-plan":      {},
	// Code
	"code-source":   {},
	"code-example":  {},
	"code-session":  {},
	"code-repl":     {},
	"code-config":   {},
	"code-diff":     {},
	"code-error":    {},
	"code-schema":   {},
	"code-template": {},
	"code-log":      {},
	"code-query":    {},
	"code-wire":     {},
	"code-build":    {},
	// Lists
	"list-toc":        {},
	"list-check":      {},
	"list-steps":      {},
	"list-faq":        {},
	"list-deps":       {},
	"list-breadcrumb": {},
	"list-related":    {},
	"list-errors":     {},
	"list-decisions":  {},
	"list-reqs":       {},
	"list-api":        {},
	"list-changelog":  {},
	// Callouts
	"callout-example":      {},
	"callout-deprecated":   {},
	"callout-security":     {},
	"callout-perf":         {},
	"callout-compat":       {},
	"callout-experimental": {},
	"callout-see-also":     {},
	"callout-breaking":     {},
	"callout-success":      {},
	"callout-quote":        {},
	// Prose
	"prose-summary":     {},
	"prose-caption":     {},
	"prose-rationale":   {},
	"prose-caveat":      {},
	"prose-migration":   {},
	"prose-changelog":   {},
	"prose-release":     {},
	"prose-prereq":      {},
	"prose-abstract":    {},
	"prose-deprecation": {},
	"prose-attribution": {},
	"prose-definition":  {},
	"prose-annotation":  {},
	"prose-example":     {},
}

// IsFutureDirective reports whether name is a declared-but-unimplemented
// directive.
func IsFutureDirective(name string) bool {
	_, ok := futureDirectives[name]
	return ok
}

// FutureDirectiveNames returns every declared-but-unimplemented directive
// name, sorted.
func FutureDirectiveNames() []string {
	out := make([]string, 0, len(futureDirectives))
	for name := range futureDirectives {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// AllBuiltinDirectives returns the set of every built-in directive name --
// the core catalogue plus the future names -- as a fresh set the caller may
// keep.
//
// This is the name set the directive parser validates against.
func AllBuiltinDirectives() map[string]struct{} {
	out := make(map[string]struct{}, Core().Len()+len(futureDirectives))
	for _, name := range Core().Names() {
		out[name] = struct{}{}
	}
	for name := range futureDirectives {
		out[name] = struct{}{}
	}
	return out
}

// AllBuiltinDirectiveNames returns every built-in directive name, sorted.
func AllBuiltinDirectiveNames() []string {
	set := AllBuiltinDirectives()
	out := make([]string, 0, len(set))
	for name := range set {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// IsBuiltinDirective reports whether name is a core or future built-in.
func IsBuiltinDirective(name string) bool {
	return Core().Has(name) || IsFutureDirective(name)
}

// IsValidDirective reports whether name is a recognized built-in or one of
// customNames. A nil customNames means the project declares no custom
// directives.
func IsValidDirective(name string, customNames map[string]struct{}) bool {
	if IsBuiltinDirective(name) {
		return true
	}
	if customNames != nil {
		if _, ok := customNames[name]; ok {
			return true
		}
	}
	return false
}

// DirectiveStatus returns "core", "future" or "unknown" for name.
//
// Only built-ins are judged; a custom directive is the caller's business and
// reads as unknown here.
func DirectiveStatus(name string) string {
	if Core().Has(name) {
		return "core"
	}
	if IsFutureDirective(name) {
		return "future"
	}
	return "unknown"
}

// DirectiveAttrError reports a directive that used an attribute it does not
// accept, or omitted one it requires.
//
// This is a hard error, distinct from a resolution failure, which is
// warning-level.
type DirectiveAttrError struct {
	Message string
}

func (e *DirectiveAttrError) Error() string { return e.Message }

// ValidateDirectiveAttrs enforces a directive's attribute contract against its
// catalogue spec.
//
// It returns a [DirectiveAttrError] when attrs carries an attribute the
// directive does not accept, or omits one it requires. Only core directives
// have a spec to enforce; custom and future directives are skipped, because
// they define their own attribute contracts.
//
// file and line name the directive's source position in the message.
func ValidateDirectiveAttrs(name string, attrs map[string]string, file string, line int) error {
	spec, ok := Core().Spec(name)
	if !ok {
		return nil
	}

	allowed := make(map[string]struct{}, len(spec.RequiredAttrs)+len(spec.OptionalAttrs))
	for _, a := range spec.RequiredAttrs {
		allowed[a] = struct{}{}
	}
	for _, a := range spec.OptionalAttrs {
		allowed[a] = struct{}{}
	}

	// The attribute order a map iteration gives is arbitrary, and the message
	// names one offending attribute, so the refusal is decided over a sorted
	// pass -- the same occurrence always reports the same attribute.
	present := make([]string, 0, len(attrs))
	for attr := range attrs {
		present = append(present, attr)
	}
	sort.Strings(present)

	for _, attr := range present {
		if _, ok := allowed[attr]; ok {
			continue
		}
		// An actionable migration note for the removed table-commands path
		// attribute.
		if name == "table-commands" && attr == "path" {
			return &DirectiveAttrError{Message: fmt.Sprintf(
				"%s:%d: directive 'table-commands' no longer takes "+
					"'path'; the schema is discovered automatically. Use "+
					`schema-dir="<dir>" only if discovery reports ambiguity.`,
				file, line)}
		}
		allowedDisplay := "(none)"
		if len(allowed) > 0 {
			sorted := make([]string, 0, len(allowed))
			for a := range allowed {
				sorted = append(sorted, a)
			}
			sort.Strings(sorted)
			allowedDisplay = strings.Join(sorted, ", ")
		}
		return &DirectiveAttrError{Message: fmt.Sprintf(
			"%s:%d: directive '%s' has unknown attribute "+
				"'%s'. Allowed attributes: %s.",
			file, line, name, attr, allowedDisplay)}
	}

	for _, req := range spec.RequiredAttrs {
		if _, ok := attrs[req]; !ok {
			return &DirectiveAttrError{Message: fmt.Sprintf(
				"%s:%d: directive '%s' is missing required "+
					"attribute '%s'. Required attributes: %s.",
				file, line, name, req, strings.Join(spec.RequiredAttrs, ", "))}
		}
	}

	return nil
}
